### PR TITLE
Always cast all inputs to given dtype in gradient check

### PR DESCRIPTION
**Pull request by hvy**
When passing the `no_grads` argument to the gradient check while at the same time specifying the `dtype` argument, all inputs are cast to the given `dtype` except the ones ignored by `no_grads`.
dtypes of function arguments are thus occasionally mixed, in particular, if a function that is being tested for correct gradients uses a kernel, an error is raises on mismatched dtypes on the kernel call.

This PR fixes this issue by always casting all inputs if `dtype` is specified no matter the value of `no_grads`.

### CODE REVIEW

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -269,8 +272,9 @@ def check_backward(func, x_data, y_grad, params=(),
         no_grads = [x.dtype.kind != 'f' for x in xs]
     else:
         if len(no_grads) != len(xs):
-            raise ValueError(
-                'Length of no_grads param and xs should be same.')
+            msg = ('Length of no_grads param and xs should be same.\n'
+                   'Actual: {0} != {1}'.format(len(no_grads), len(xs)))
+            raise ValueError(msg)
```
Comment (niboshi) [inline chainer/gradient_check.py]: You don't have to assign a temporary variable here.

File: tests/chainer_tests/test_gradient_check.py [hunk 2]
```diff
@@ -358,7 +358,7 @@ def f(x, y):
 
     def test_no_grads_option(self):
         x1 = numpy.array([1], dtype='f')
-        x2 = numpy.array([1], dtype='f')  # grad check for this is skipped
+        x2 = numpy.array([1], dtype='f')
```
Comment (niboshi) [inline tests/chainer_tests/test_gradient_check.py]: Why did you remove the comment?

File: chainer/gradient_check.py (hunk 1 above)
Comment (hvy) [reply chainer/gradient_check.py]: Thanks, I removed the redundant variable.

File: tests/chainer_tests/test_gradient_check.py (hunk 2 above)
Comment (hvy) [reply tests/chainer_tests/test_gradient_check.py]: Fixed.

**Comment by niboshi** [comment]
Please fix conflict.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by niboshi** [comment]
LGTM!
